### ISSUE TITLE
The ROS node error when inference rosbag data

### ISSUE THREAD

**Issue by muzi2045** [cedricxie/second_ros#2]
![image](https://user-images.githubusercontent.com/12984898/48670381-11022900-eb52-11e8-8b95-222b0b3305a6.png)
the second node can launch with no error, But when the rosbag was played, error occured like below.

![image](https://user-images.githubusercontent.com/12984898/48670402-5fafc300-eb52-11e8-8cd1-dd35032dc53d.png)

Ubuntu 16.04 LTS
ROS-kinect
python 3.6.7
pytorch 0.4.1

the rosbag data is convert from raw data in kitti 2011.09.26 0005 by kitti2bag tool.
hopefully for any advice, thanks !

**Comment by cedricxie** [comment]
Hi @muzi2045 , could you please check PYTHONPATH and see if you have correctly set it? It looks like the "original" TourchInferenceContext object is called and therefore it is missing the function I added later.

Yuesong

**Comment by muzi2045** [comment]
thanks a lot, I have run it successfully , But obviously the detect result is wrong , I am trying to figure it out

**Comment by pangsu0613** [comment]
@muzi2045 hello, could you please tell me how you solve this problem? I got the same error. Thank you so much.

**Comment by muzi2045** [comment]
check your PYTHONPATH, maybe you using the wrong object TorchInferenceContext, because the author modify this object and add a new funtion, check it and set correct PYTHONPATH.

**Comment by pangsu0613** [comment]
@muzi2045 Thank you so much for the quick response, it is my PYTHONPATH problem.

**Comment by mzhian** [comment]
> thanks a lot, I have run it successfully , But obviously the detect result is wrong , I am trying to figure it out
@muzi2045 

Hi muzi, I also find the detect result is wrong, please refer to the issue #4, where I posted.  I tried both the pretrained models and the new model I trained myself, both the same issue. 
Have you figured out what went wrong?
Thanks a lot!

**Comment by muzi2045** [comment]
I guess the reason is the detecting time per frame cost too much ms , so you can't show the detecting bounding box realtime~~

**Comment by mzhian** [comment]
@muzi2045

Yeah you are right, when i play the bag slowly It seems working well then. I'll see where in the code costs so much time, 'cause the second model is supposed to be realtime. Thank you so much.

**Comment by muzi2045** [comment]
@mzhian  you can try this https://github.com/nutonomy/second.pytorch , modify it to ROS, this repo infer speed is faster than origin repo, but it also can't realtime, maybe 16-line Lidar data can reach to 10HZ

**Comment by pangsu0613** [comment]
@muzi2045 @mzhian There are mainly two reasons for the processing delay, I changed two positions and can achieve 10hz on kitti rosbags.
1. There are some problems in the ROS python subscriber which makes it slow in reading data from rosbag. Adding a buff_size = 2^24 in the rospy subscriber will fix this issue.
2. In the original code, the way that the author publishing modified point cloud is very slow, the function that the author uses to generate Pointcloud2 messages are very slow. Change another way or just delete that part. publishing modified point cloud is not useful for the detection. Delete "publish_test(proc.inputs['points'][0], msg.header.frame_id)".

**Comment by muzi2045** [comment]
@pangsu0613  
 youm mean change the queue_size to 2^24 in the following code?
`sub_ = rospy.Subscriber("/kitti/velo/pointcloud", PointCloud2, rslidar_callback, queue_size=1)`

**Comment by pangsu0613** [comment]
@muzi2045 no, add "buff_size = 2**24" after queue_size = 1

**Comment by muzi2045** [comment]
@pangsu0613 
it looks better when add buff_size = 2**24, but it can't solve the core problem
this is my test inference time cost
`Start inference_vxnet...
input preparation time: 0.02529454231262207
convert_time:  0.016423463821411133
example_convert_to_torch_time:  0.017003297805786133
predict_time:  0.10981249809265137
detection time: 0.12741827964782715`

obviously the detection time and dataset convert time both cost too much.

**Comment by pangsu0613** [comment]
@muzi2045 delete publish_test(proc.inputs['points'][0], msg.header.frame_id) would also help. My PC has NVIDIA 1080 Ti, after doing these two steps. I can achieve 10 Hz. I am not at lab now, I can not show you the exact detection time. But I can recall for sure that my detection time is just around 0.04s. Another thing I have changed is that I use SECOND V1.5.

**Comment by muzi2045** [comment]
1080Ti ....
OK then, I only has 1050Ti...., I'll try run it in Jetson Xavier, maybe it's better than my PC

**Comment by pangsu0613** [comment]
@muzi2045 Ok. Even with 1080 Ti, I can only achieve around 4~5 hz before I made these changes. But the detection time at that time was still shorter than you. I think running it on a better GPU would be a good choice. Good luck.

**Comment by mzhian** [comment]
@muzi2045 @pangsu0613 
Thank both of you for your suggestions. I've added the buffer size and removed publishing the converted pointclouds, it now can reach 10 hz detection rate using the Second V1 model. I'll also try out the nutonomy's repo as well. 
Great job and good luck!

**Comment by mzhian** [comment]
@muzi2045 @pangsu0613 
FYI, I found another place where drags down the speed (at least for my case):
In the `velo_callback`, there is a point coords extration step to convert the msg to np arrays: 
`pcl_msg = pc2.read_points(msg, skip_nans=False, field_names=("x", "y", "z","intensity","ring"))`
`np_p = np.array(list(pcl_msg), dtype=np.float32)`
this part takes even longer time than the inference time, which costs around `140` ms for around `120,000` points. I'm not sure if it is just for my case or not, but after I replaced the operation with the [ros_numpy.point_cloud2](http://docs.ros.org/kinetic/api/ros_numpy/html/point__cloud2_8py_source.html) function: 
`xyz_array = ros_numpy.point_cloud2.pointcloud2_to_xyz_array(msg)`
it now takes only `1.8` ms to extract the xyz readings. See my terminal output
![screenshot from 2019-02-20 13-50-23](https://user-images.githubusercontent.com/22515441/53071733-1179d980-351e-11e9-8ff1-d91823b663ac.png)
The `ros_numpy.point_cloud2` library seems to offer much faster conversion from(to) pointcloud to(from) numpy array as I tested, so probably the `publish_test ` can also be rewritten with this library to create pointcloud from the points instead of using the `pc2.create_cloud`.

**Comment by pangsu0613** [comment]
@mzhian Thank you for your reminding. I tried your method, it does increase the speed. and yes, rewriting the publish_test using your new ros_numpy function is gonna make it faster. But I think there is no need to run publish_test, it does not contribute anything to the inferance, that's just something the author wrote for some early stage debug I think.

**Comment by mzhian** [comment]
@pangsu0613 
Great to know it works. Credit to the `ros_numpy.point_cloud2 `author `Jon Binney` from Willow Garage 
And yes I agree the publish_test should be for debug use as commented by the author. For my case I'm actually using that to visualize the bounding boxes (only for visualization purposes), as it is synchronized with the detection boxes, otherwise the boxes will be a bit out-of-phase with the pointcloud if using the pointcloud topic from the rosbag, due to the detection delay.    
BTW, this repo is very helpful to begin with the integration work into ROS, Let's appreciate the author for sharing it!
Cheers.

**Comment by muzi2045** [comment]
@cedricxie @pangsu0613 
There has  another question about how to inference the all pointcloud data in KITTI and output the boundingbox, Not just output the front view of the car in the pointcloud.
Thanks.

**Comment by mzhian** [comment]
@muzi2045 
Hi Muzi, the solution can be found through issues #12 and #65. 
The 1st approach is to just adjust the range parameter settings in the `.config` files, to make x < 0 for `point_cloud_range`, `post_center_limit_range`, and `anchor_ranges`. But the pretrained models are trained on the front views only, I tried it and found the result not good. To achieve good performance for x < 0 you'd better train the model yourself using the 360 view of the pointclouds.
A 2nd approach is to run two inference models together, one for the front, one for the back. Doing so you need to rotate the input pointcloud 180 degrees for the 2nd inference, and rotate 180 degrees back the detection bboxes. Merging the bboxes from the two inferences will give you the 360 degree detection (make sure the bboxes are using the same `frame_id`). I tried this also and it works pretty well on the pretrained models.

**Comment by muzi2045** [comment]
I found the most cost time part in the network forward is the predict. 
Which locate at the boxs_preds[mask],
In pytorch, when number in the Tensor reach to 100K, the boolean index slice will cost 40ms, maybe longer if you input more pointcloud.

**Comment by Tom698** [comment]
> @muzi2045 Hi Muzi, the solution can be found through issues #12 and #65. The 1st approach is to just adjust the range parameter settings in the `.config` files, to make x < 0 for `point_cloud_range`, `post_center_limit_range`, and `anchor_ranges`. But the pretrained models are trained on the front views only, I tried it and found the result not good. To achieve good performance for x < 0 you'd better train the model yourself using the 360 view of the pointclouds. A 2nd approach is to run two inference models together, one for the front, one for the back. Doing so you need to rotate the input pointcloud 180 degrees for the 2nd inference, and rotate 180 degrees back the detection bboxes. Merging the bboxes from the two inferences will give you the 360 degree detection (make sure the bboxes are using the same `frame_id`). I tried this also and it works pretty well on the pretrained models.

Thank you for your advice Mzhian! I want to konw how to rotate the input pointcloud and I have tried multiple times without success. Would you provide some suggestions please?
